### ISSUE TITLE
Out of bound pointer causing DTLS server to crash

### ISSUE THREAD

**Issue by bathooman** [openssl/openssl#14906]
Version Affected: openssl-3.0.0-alpha12
Configuration Options: `CC=gcc ./Configure --debug --strict-warnings no-asm no-shared no-threads no-tests -no-dso enable-ssl-trace`

Hello,

This bug occurs when the server receives `Finished` as the first message. This bug is found by employing symbolic execution technique. Suppose, we send the following record to the DTLS server:

![image](https://user-images.githubusercontent.com/3464382/115112665-da30d700-9f86-11eb-85c6-b4e750cdb9a5.png)

Note that the handshake message is indeed a `Client Hello`. I just changed the `Handshake Type` to `Finished`. When this message is received by the server, the server goes for MAC computation. Inside `tls1_final_finish_mac ()` at this [point](https://github.com/openssl/openssl/blob/openssl-3.0.0-alpha12/ssl/t1_enc.c#L617) since there is no cipher suite selected previously, we have an out-of-bound pointer that crashes the server.
The stack trace for where this bug happens is as the following:

```
tls1_final_finish_mac (=94554126172160, =94554414924736, =15, =94554126172744) at ssl/t1_enc.c:617
ssl3_take_mac (=94554126172160) at ssl/statem/statem_lib.c:719
in dtls_get_message (=94554126172160, =94554491407848, =94554491408432) at ssl/statem/statem_dtls.c:385
read_state_machine (=94554126172160) at ssl/statem/statem.c:580
state_machine (=94554126172160, =1) at ssl/statem/statem.c:437
in ossl_statem_accept (=94554126172160) at ssl/statem/statem.c:265
SSL_do_handshake (=94554126172160) at ssl/ssl_lib.c:3859

```
I have attached the means to reproduce the bug. To do so, after compiling OpenSSL-3.0.0-alpha12, start a DTLS server using the OpenSSL utility in `openssl-openssl-3.0.0-alpha12/apps`:

`./openssl s_server -dtls`

The DTLS server starts listening on port 4433. Now, by executing `reproduce.sh` in `openssl-openssl-3.0.0-alpha12/bug-reproduce` directory, you can cause the server to crash like the following:

![image](https://user-images.githubusercontent.com/3464382/115113307-f84c0680-9f89-11eb-9b35-744b14008cca.png)

`reproduce.sh` solely sends the mentioned `Finished` message to the server using `nc` utility.

[openssl-openssl-3.0.0-alpha12.zip](https://github.com/openssl/openssl/files/6329239/openssl-openssl-3.0.0-alpha12.zip)

**Update**: The bug also affects the latest version: `openssl-3.0.0-alpha14`

**Comment by t8m** [comment]
@mattcaswell could you please look at this?

**Comment by mattcaswell** [comment]
Good catch! Fortunately this only affects the dev branch and not 1.1.1 (otherwise this would have been a CVE).

Fix for this in #14930.
